libpretixui-android/src/main/java/eu/pretix/libpretixui/android/scanning/ScannerView.kt

```diff
@@ -1,30 +1,298 @@
+import androidx.camera.core.Camera
+import androidx.camera.core.CameraSelector
+import androidx.camera.core.ImageAnalysis
+import androidx.camera.core.ImageProxy
+import androidx.camera.core.ImageProxy.PlaneProxy
+import androidx.camera.core.Preview
+import androidx.camera.core.TorchState
+import androidx.camera.lifecycle.ProcessCameraProvider
+import androidx.camera.view.PreviewView
+import androidx.core.content.ContextCompat
+import androidx.lifecycle.findViewTreeLifecycleOwner
+import com.google.zxing.BinaryBitmap
+import com.google.zxing.MultiFormatReader
+import com.google.zxing.NotFoundException
+import com.google.zxing.PlanarYUVLuminanceSource
+import com.google.zxing.common.HybridBinarizer
+import java.lang.IllegalArgumentException
+import java.nio.ByteBuffer
+import java.util.concurrent.ExecutorService
+import java.util.concurrent.Executors
+import java.util.concurrent.atomic.AtomicBoolean
 
 
-class ScannerView : ZXingScannerView {
-    private var mFramingRectInPreview: Rect? = null
+@SuppressLint("UnsafeOptInUsageError")
+class ScannerView : FrameLayout {
+    data class Result(
+        val text: String,
+        val rawBytes: ByteArray,
+    )
+
+    interface ResultHandler {
+        fun handleResult(rawResult: Result)
+    }
+
+    private var resultHandler: ResultHandler? = null
+    private var cameraProvider: ProcessCameraProvider? = null
+    private var cameraExecutor: ExecutorService? = null
+    private var previewView: PreviewView? = null
+    private var torchState: Boolean = false
+    private var torchTarget: Boolean = false
+    private var autofocusState: Boolean = false
+    private var autofocusTarget: Boolean = false
+    private var camera: Camera? = null
 
     constructor(context: Context) : super(context) {}
 
     constructor(context: Context, attributeSet: AttributeSet) : super(context, attributeSet) {}
 
-    @Synchronized
-    override fun getFramingRectInPreview(previewWidth: Int, previewHeight: Int): Rect {
-        if (this.mFramingRectInPreview == null) {
-            val rect = Rect()
-            rect.left = 0
-            rect.top = 0
-            rect.right = previewWidth
-            rect.bottom = previewHeight
+    var torch: Boolean
+        get() = torchTarget
+        set(value) {
+            torchTarget = value
+            if (torchState != torchTarget) {
+                camera?.cameraControl?.enableTorch(torchTarget)
+            }
+        }
+
+    var autofocus: Boolean
+        get() = autofocusTarget
+        set(value) {
+            autofocusTarget = value
+            if (autofocusTarget != autofocusState) {
+                enableAutofocus(autofocusTarget)
+            }
+        }
+
+    override fun addView(child: View?) {
+        super.addView(child)
+    }
+
+    fun setResultHandler(rh: ResultHandler) {
+        this.resultHandler = rh
+    }
+
+    fun startCamera() {
+        removeAllViews()
+        previewView = PreviewView(context)
```

Comment:
this uses implicit `implementationMode = PreviewView.ImplementationMode.PERFORMANCE`, which sadly breaks ui rendering on my test devices (led-tablet A10, sunmi v3mix A13).
Even setting the `FLAG_HARDWARE_ACCELERATED` on the containing window, as soon as the camera image is rendered, the background of the surrounding elements goes missing. Thats especially bad, when a textbox for manual entry should be visible to the user.

The only fix for this is currently `implementationMode = PreviewView.ImplementationMode.COMPATIBLE`, for which the documentation says that performance will be not great, but thats something I couldn't confirm on my devices.